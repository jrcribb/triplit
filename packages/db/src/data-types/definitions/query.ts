import { CollectionNameFromModels } from '../../db.js';
import {
  CollectionNotFoundError,
  InvalidQueryCardinalityError,
  TriplitError,
} from '../../errors.js';
import { Models } from '../../schema/types/index.js';
import { TypeInterface } from './type.js';
import {
  CollectionQuery,
  FetchResult,
  QueryResultCardinality,
} from '../../query/types/index.js';

export type SubQuery<
  M extends Models,
  CN extends CollectionNameFromModels<M>
> = Pick<
  CollectionQuery<M, CN, any, any>,
  'collectionName' | 'where' | 'limit' | 'order'
>;

// In practice, we are able to infer the collection name from the query Q['collectionName'] (not CN)
export type QueryType<
  CN extends CollectionNameFromModels,
  Q extends SubQuery<Models, CN>,
  C extends QueryResultCardinality = 'many'
> = TypeInterface<
  'query',
  FetchResult<Models, Q>,
  any, //TODO: is this even applicable? ... might need to break it out into its own concepts we slowly add to
  readonly []
> & {
  query: Q;
  cardinality: C;
};

export function QueryType<
  CN extends CollectionNameFromModels,
  Q extends SubQuery<Models, CN>,
  C extends QueryResultCardinality = 'many'
>(query: Q, cardinality: C = 'many' as C): QueryType<CN, Q, C> {
  return {
    type: 'query' as const,
    supportedOperations: [] as const, // 'hasKey', etc
    context: {},
    cardinality,
    query,
    toJSON() {
      // TODO verify this works with non-memory storage providers
      return { type: this.type, query, cardinality };
    },
    convertInputToDBValue(val: any) {
      // TODO: this is a placeholder, this could be a place where we could support insertions with relationships
      throw new TriplitError(
        'Invalid Operation - Inserting relational data is not supported'
      );
    },
    // There isnt exactly a "DB value" for queries, but this makes sense to use with the fetch method
    convertDBValueToJS(val, schema) {
      if (!schema)
        throw new TriplitError(
          'A schema is required to convert DB value to JS'
        );
      const relationSchema = schema?.[query.collectionName]?.schema;
      if (!relationSchema)
        throw new CollectionNotFoundError(query.collectionName, schema);

      // TODO: determine when we would see this
      if (!val) return val;

      if (cardinality === 'one') {
        return relationSchema.convertDBValueToJS(val, schema);
      } else if (cardinality === 'many') {
        // TODO: confirm map
        // if (val === undefined) return undefined;
        // const entries = JSON.parse(val);
        return Array.from(val.entries()).map(([, v]: any) =>
          relationSchema.convertDBValueToJS(v, schema)
        );
      } else {
        throw new InvalidQueryCardinalityError(cardinality);
      }
    },
    convertJSONToJS(val, schema) {
      if (!schema)
        throw new TriplitError(
          'A schema is required to convert JSON value to JS'
        );
      const relationSchema = schema?.[query.collectionName]?.schema;
      if (!relationSchema)
        throw new CollectionNotFoundError(query.collectionName, schema);

      if (!val) return val;
      if (cardinality === 'one') {
        return relationSchema.convertJSONToJS(val, schema);
      } else if (cardinality === 'many') {
        return val.map((v: any) => relationSchema.convertJSONToJS(v, schema));
      } else {
        throw new InvalidQueryCardinalityError(cardinality);
      }
    },
    convertJSToJSON(val, schema) {
      if (!schema)
        throw new TriplitError(
          'A schema is required to convert JSON value to JS'
        );
      const relationSchema = schema?.[query.collectionName]?.schema;
      if (!relationSchema)
        throw new CollectionNotFoundError(query.collectionName, schema);

      if (!val) return val;

      if (cardinality === 'one') {
        return relationSchema.convertJSToJSON(
          // @ts-expect-error Need to fixup type to understand that this is a single value
          val,
          schema
        );
      } else if (cardinality === 'many') {
        return val.map((v: any) => relationSchema.convertJSToJSON(v, schema));
      } else {
        throw new InvalidQueryCardinalityError(cardinality);
      }
    },
    defaultInput() {
      return undefined;
    },
    validateInput(_val: any) {
      return undefined; // TODO
    },
    validateTripleValue(_val: any) {
      return false;
    },
  };
}
