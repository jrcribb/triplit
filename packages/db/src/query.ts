import { hashObject } from './utils/hash.js';
import { CollectionNameFromModels, Models } from './schema/types/models.js';
import { StringKey, Unalias } from './utils/types.js';
import {
  Decoded,
  ModelPaths,
  ModelRelationshipPaths,
  PathFiltered,
  ResolveRelationshipPath,
  SchemaPaths,
} from './schema/index.js';

/**
 * Hashes a query object to a unique string, ignoring non-query properties. Thus the hash is of the query the server will see.
 */
export function hashQuery<Q extends CollectionQuery>(params: Q) {
  const queryParams = Object.fromEntries(
    Object.entries(params).filter(([key]) =>
      (COLLECTION_QUERY_PROPS as string[]).includes(key)
    )
  );
  const hash = hashObject(queryParams).toString(); // Hash(queryParams).toString();
  return hash;
}

// Should be friendly types that we pass into queries
// Not to be confused with the Value type that we store in the triple store
// TODO: this could be more speciifc
export type QueryValue =
  | number
  | string
  | boolean
  | Date
  | null
  | number[]
  | boolean[]
  | string[]
  | Date[];

// // TODO: rename
// Record of all collection queries
export type SchemaQueries<M extends Models<M> = Models> = {
  [CN in CollectionNameFromModels<M>]: CollectionQuery<M, CN>;
};

// Union of all collection queries if CN not provided
export type SchemaQuery<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = SchemaQueries<M>[CN];

export interface CollectionQuery<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> {
  collectionName: CN;
  select?: QuerySelection<M, CN>[];
  where?: QueryWhere<M, CN>;
  order?: QueryOrder<M, CN>;
  limit?: number;
  after?: QueryAfter;
  vars?: Record<string, any>;
  // TODO: add back inclusions
  include?: QueryInclusions<M, CN>;
  // /**
  //  * @deprecated define a where filter instead
  //  */
  // entityId?: string; // Syntactic sugar for where("id", "=", entityId), should not be relied on in query engine
}

/**
 * Cardinality of a query result:
 * - 'one' - a single result
 * - 'many' - multiple results
 */
export type QueryResultCardinality = 'one' | 'many';

export type QuerySelectionFromQuery<Q extends CollectionQuery<any, any>> =
  Q extends CollectionQuery<infer M, infer CN> ? QuerySelection<M, CN> : never;
export type QuerySelection<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = SchemaPaths<M, CN>;

// ====== Filter Types ======
/**
 * A query filter, which is a collection of many filters.
 */
export type QueryWhere<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = WhereFilter<M, CN>[];

/**
 * A single filter, which may have various structures.
 */
// I've done this with ExistsFilter, but adding a 'type' property to each type for narrowing would be helpful. Should still support old props for backwards compatibility.
export type WhereFilter<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> =
  // TODO: investigate usage of ModelFilterStatements (may improve type completion)
  | FilterStatement<M, CN>
  | FilterGroup<M, CN>
  | SubQueryFilter<M>
  | RelationshipExistsFilter<M, CN>
  | boolean;

export type ModelFilterStatements<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = {
  [K in ModelPaths<M, CN>]: FilterStatement<M, CN, K>;
};
export type ModelFilterStatement<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
  K extends ModelPaths<M, CN> = ModelPaths<M, CN>,
> = ModelFilterStatements<M, CN>[K];

/**
 * A single filter statement of the shape [path, operator, value].
 */
export type FilterStatement<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
  K extends ModelPaths<M, CN> = ModelPaths<M, CN>,
> = [
  K,
  //Operations<ResolveModelPath<M, CN, K>> I think typescript has trouble inferring this because its a tuple, seeing a union of all ops, which is fine for now
  // However, it also causes small issues with the type system with ex. try S.RelationMany('users', { where: [ or([ ['id', '=', '$liked_by_ids'], or([['id', '=', '$liked_by_ids']]), ]), ], })
  // It (i think) doesnt recognize string in ['id', string, string] as a valid operator and is failing
  string,
  any,
];

/**
 * A set of filters specified to be combined with AND or OR.
 */
export type FilterGroup<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = AndFilterGroup<M, CN> | OrFilterGroup<M, CN>;

/**
 * A group of filters combined with AND.
 */
export type AndFilterGroup<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = {
  mod: 'and';
  filters: QueryWhere<M, CN>;
};

/**
 * A group of filters combined with OR.
 */
export type OrFilterGroup<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = {
  mod: 'or';
  filters: QueryWhere<M, CN>;
};

/**
 * An exists filter that will check if a subquery returns any results.
 */
export type SubQueryFilter<
  M extends Models<M> = Models,
  SQ extends SchemaQuery<M> = SchemaQuery<M>,
  // This is the collection name of the subquery, not the parent query
  // SubqueryCN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = {
  exists: SQ;
};

/**
 * An exists filter that will check if a relationship in the schema returns any results.
 */
// This may be never if M = any ... might be a place to be more flexible
export type RelationshipRef<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = StringKey<NonNullable<M[CN]['relationships']>>;

export type IsRelationshipRef<
  M extends Models<M>,
  CN extends CollectionNameFromModels<M>,
  Ref extends string,
> = Ref extends RelationshipRef<M, CN> ? true : false;

// TODO: unify with { exists }

export type RelationshipExistsFilter<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
  P extends ModelRelationshipPaths<M, CN> = ModelRelationshipPaths<M, CN>,
  Ext extends RelationshipExistsExtension<
    M,
    CN,
    P
  > = RelationshipExistsExtension<M, CN, P>,
> = {
  exists: Ext & {
    _extends: P;
  };
};

export type RelationshipExistsExtension<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
  P extends ModelRelationshipPaths<M, CN> = ModelRelationshipPaths<M, CN>,
> = Pick<
  SchemaQuery<M, ResolveRelationshipPath<M, CN, P>['query']['collectionName']>,
  'where'
>;

// ====== Order Types ======
/**
 * A query order, which is a collection of many orders.
 */
export type QueryOrder<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = OrderStatement<M, CN>[];

/**
 * A single order statement of the shape [path, direction].
 */
export type OrderStatement<
  M extends Models<M>,
  CN extends CollectionNameFromModels<M>,
> = [property: ModelPaths<M, CN>, direction: 'ASC' | 'DESC'];

// ====== Pagination Types ======
export type QueryAfter = [ValueCursor, boolean];
export type ValueCursor = [value: QueryValue, ...values: QueryValue[]];

// ====== Inclusion Types ======
/**
 * A map of inclusions, keyed by alias.
 */
export type QueryInclusions<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = {
  // Optimally we could exclude RefShorthand from non-relationship keys, but this is tricky to do with TS
  [K in string]: QueryInclusion<M, CN>;
};

/**
 * A possible inclusion value in a query.
 */
export type QueryInclusion<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
> = RefShorthand | RelationSubquery<M> | RefSubquery<M, CN>;

/**
 * A shorthand for including a reference.
 */
export type RefShorthand = true | null;

// ========= Ref Subquery Types =========
export type RefSubquery<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
  Ref extends RelationshipRef<M, CN> = RelationshipRef<M, CN>,
> = {
  _extends: Ref;
} & RefQueryExtension<M, CN, Ref>;

/**
 * An extension of a referential subquery, specifying additional query parameters.
 */
export type RefQueryExtension<
  M extends Models<M> = Models,
  CN extends CollectionNameFromModels<M> = CollectionNameFromModels<M>,
  Ref extends RelationshipRef<M, CN> = RelationshipRef<M, CN>,
> = Pick<
  SchemaQuery<M, RefCollectionName<M, CN, Ref>>,
  'select' | 'include' | 'limit' | 'where' | 'order'
>;

export type RefDefinition<
  M extends Models<M>,
  CN extends CollectionNameFromModels<M>,
  Ref extends RelationshipRef<M, CN>,
> = NonNullable<M[CN]['relationships']>[Ref];

export type RefQuery<
  M extends Models<M>,
  CN extends CollectionNameFromModels<M>,
  Ref extends RelationshipRef<M, CN>,
> = RefDefinition<M, CN, Ref>['query'];

export type RefCollectionName<
  M extends Models<M>,
  CN extends CollectionNameFromModels<M>,
  Ref extends RelationshipRef<M, CN>,
> = RefQuery<M, CN, Ref>['collectionName'];

// ========= Relational Subquery Types =========
/**
 * A subquery defining a relationship, specifying the subquery and cardinality of the result.
 */
export type RelationSubquery<
  M extends Models<M> = Models,
  Q extends SchemaQuery<M> = SchemaQuery<M>,
  Cardinality extends QueryResultCardinality = QueryResultCardinality,
> = {
  subquery: Q;
  cardinality: Cardinality;
};

export type FetchResult<
  M extends Models<M>,
  Q extends SchemaQuery<M>,
  C extends QueryResultCardinality,
> = Unalias<AliasedFetchResult<M, Q, C>>;

export type AliasedFetchResult<
  M extends Models<M>,
  Q extends SchemaQuery<M>,
  C extends QueryResultCardinality,
> = C extends 'one'
  ? AliasedQueryResult<M, Q> | null
  : AliasedQueryResult<M, Q>[];

type AliasedQueryResult<M extends Models<M>, Q extends SchemaQuery<M>> =
  Q extends CollectionQuery<M, infer CN>
    ? (Q['select'] extends ReadonlyArray<infer S>
        ? // If we have a selection, use that
          PathFiltered<Decoded<M[CN]['schema']>, S extends string ? S : never>
        : // Else use the entire schema
          Decoded<M[CN]['schema']>) & {
        // also use inclusions
        [K in StringKey<Q['include']>]: InclusionResult<
          M,
          CN,
          K,
          // @ts-expect-error
          Q['include'][K]
        >;
      }
    : never;

type InclusionResult<
  M extends Models<M>,
  CN extends CollectionNameFromModels<M>,
  Alias extends string,
  Inclusion extends QueryInclusion<M, CN>,
> =
  // Inclusion is subquery
  Inclusion extends RelationSubquery<M, infer Subquery, infer Cardinality>
    ? AliasedFetchResult<M, Subquery, Cardinality>
    : // Inclusion is relation extension
      Inclusion extends RefSubquery<M, CN, infer Ref>
      ? AliasedFetchResult<
          M,
          RefQuery<M, CN, Ref> & Omit<Inclusion, '_extends'>,
          RefDefinition<M, CN, Ref>['cardinality']
        >
      : // Inclusion is relation shorthand
        Inclusion extends RefShorthand
        ? Alias extends RelationshipRef<M, CN>
          ? AliasedFetchResult<
              M,
              RefDefinition<M, CN, Alias>['query'],
              RefDefinition<M, CN, Alias>['cardinality']
            >
          : never
        : never;

// Should update this as we add more query properties
const COLLECTION_QUERY_PROPS = [
  'after',
  'collectionName',
  'select',
  // 'entityId',
  'include',
  'limit',
  'order',
  'vars',
  'where',
] as const satisfies (keyof CollectionQuery)[];

export type WithSelection<
  Q extends CollectionQuery<any, any>,
  Selection extends QuerySelectionFromQuery<Q>,
> = Omit<Q, 'select'> & {
  select: Selection[];
};

export type WithInclusion<
  Q extends CollectionQuery<any, any>,
  Inclusion extends Q extends CollectionQuery<infer M, infer CN>
    ? QueryInclusions<M, CN>
    : never,
> = Omit<Q, 'include'> & {
  include: Inclusion;
};

export type WithInclusionRaw<Q, Inclusion> = Omit<Q, 'include'> & {
  include: Inclusion;
};
