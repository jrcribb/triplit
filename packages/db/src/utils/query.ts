import { Value } from '@sinclair/typebox/value';
import { type CollectionQuery } from '../query/types/index.js';

// Should update this as we add more query properties
const COLLECTION_QUERY_PROPS: (keyof CollectionQuery)[] = [
  'after',
  'collectionName',
  'entityId',
  'include',
  'limit',
  'order',
  'vars',
  'where',
];

/**
 * Hashes a query object to a unique string, ignoring non-query properties. Thus the hash is of the query the server will see.
 */
export function hashQuery<Q extends CollectionQuery>(params: Q) {
  const queryParams = Object.fromEntries(
    Object.entries(params).filter(([key]) =>
      (COLLECTION_QUERY_PROPS as string[]).includes(key)
    )
  );
  const hash = Value.Hash(queryParams).toString();
  return hash;
}

export function hash(value: unknown): string {
  return Value.Hash(value).toString();
}
