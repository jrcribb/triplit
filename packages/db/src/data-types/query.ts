import { FetchResult } from '../collection-query.js';
import { CollectionNameFromModels } from '../db.js';
import { CollectionQuery } from '../query.js';
import { Models } from '../schema.js';
import { TypeInterface } from './type.js';

export type SubQuery<
  M extends Models<any, any>,
  CN extends CollectionNameFromModels<M>
> = Pick<
  CollectionQuery<M, CN>,
  'collectionName' | 'where' | 'limit' | 'order'
>;

export type QueryResultCardinality = 'one' | 'many';

export type QueryType<
  Query extends SubQuery<any, any>,
  C extends QueryResultCardinality
> = TypeInterface<
  'query',
  FetchResult<Query>,
  any, //TODO: is this even applicable? ... might need to break it out into its own concepts we slowly add to
  readonly []
> & {
  query: Query;
  cardinality: C;
};

export function QueryType<
  Q extends SubQuery<any, any>,
  C extends QueryResultCardinality
>(query: Q, cardinality: C = 'many' as C): QueryType<Q, C> {
  return {
    type: 'query' as const,
    cardinality,
    supportedOperations: [] as const, // 'hasKey', etc
    query,
    toJSON() {
      // TODO verify this works with non-memory storage providers
      return { type: this.type, query, cardinality };
    },
    convertInputToDBValue(val: any) {
      return JSON.stringify(val);
    },
    convertDBValueToJS(val) {
      return val as FetchResult<Q>;
    },
    convertJSONToJS(val) {
      throw new Error('Not implemented');
    },
    convertJSToJSON(val) {
      if (!val) return val;
      // Serialize data, cardinality could be one or many
      if (cardinality === 'one') return val;
      return Array.from(val.entries());
    },
    // TODO: determine proper value and type here
    // Type should go extract the deserialized type of each of its keys
    defaultInput() {
      return undefined;
    },
    validateInput(_val: any) {
      return undefined; // TODO
    },
    validateTripleValue(_val: any) {
      return true; // TODO
    },
  };
}
