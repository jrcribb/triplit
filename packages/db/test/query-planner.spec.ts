import { describe, expect, test } from 'vitest';
import {
  compileQuery,
  compileRelationalPlan,
  extractViews,
} from '../src/query-planner/query-compiler.js';
import { CollectionQuery } from '../src/query.js';

describe('query planning', async () => {
  const queries: {
    description: string;
    query: CollectionQuery;
  }[] = [
    {
      description: "Posts authored by user with username 'Bob99'",
      query: {
        collectionName: 'posts',
        where: [
          {
            exists: {
              collectionName: 'users',
              where: [
                ['id', '=', '$1.authorId'],
                ['username', '=', 'Bob99'],
              ],
            },
          },
        ],
      },
    },
    {
      description: 'messages from conversations that the user is part of',
      query: {
        collectionName: 'messages',
        where: [
          {
            exists: {
              collectionName: 'conversations',
              where: [
                ['id', '=', '$1.conversationId'],
                {
                  exists: {
                    collectionName: 'conversation_members',
                    where: [
                      ['conversationId', '=', '$1.id'],
                      ['userId', '=', 'user123'],
                    ],
                  },
                },
              ],
            },
          },
        ],
      },
    },
    {
      description: 'Conversations including the most recent message and sender',
      query: {
        collectionName: 'conversations',
        include: {
          latestMessage: {
            subquery: {
              collectionName: 'messages',
              where: [['conversationId', '=', '$1.id']],
              order: [['timestamp', 'DESC']],
              include: {
                sender: {
                  subquery: {
                    collectionName: 'users',
                    where: [['id', '=', '$1.senderId']],
                  },
                  cardinality: 'one',
                },
              },
            },
            cardinality: 'one',
          },
        },
      },
    },
    {
      description: 'Messages and the reactions',
      query: {
        collectionName: 'messages',
        include: {
          reactions: {
            subquery: {
              collectionName: 'reactions',
              where: [['messageId', '=', '$1.id']],
            },
            cardinality: 'many',
          },
        },
      },
    },
    {
      description: '20 most recent messages in a conversation',
      query: {
        collectionName: 'messages',
        where: [['conversationId', '=', 'conv123']],
        order: [['timestamp', 'DESC']],
        limit: 20,
      },
    },
    {
      description: 'Flights to california including the destination airports',
      query: {
        collectionName: 'flights',
        where: [
          {
            exists: {
              collectionName: 'airports',
              where: [
                ['id', '=', '$1.destination'],
                ['state', '=', 'CA'],
              ],
            },
          },
        ],
        include: {
          destinationAirport: {
            subquery: {
              collectionName: 'airports',
              where: [
                ['id', '=', '$1.destination'],
                ['state', '=', 'CA'],
              ],
            },
            cardinality: 'one',
          },
        },
      },
    },
    {
      description: 'Benchmarks query',
      query: {
        collectionName: 'branches',
        include: {
          latest_run: {
            subquery: {
              collectionName: 'runs',
              select: ['created_at'],
              order: [['created_at', 'DESC']],
              where: [['branch_name', '=', '$1.id']],
            },
            cardinality: 'one',
          },
          benchmarks: {
            subquery: {
              collectionName: 'benchmarks',
              select: ['id', 'name'],
              include: {
                latest_branch_run: {
                  subquery: {
                    collectionName: 'runs',
                    select: ['results', 'created_at'],
                    order: [['created_at', 'DESC']],
                    where: [
                      ['benchmark', '=', '$1.id'],
                      ['branch_name', '=', '$2.id'],
                    ],
                  },
                  cardinality: 'one',
                },
              },
            },
            cardinality: 'many',
          },
        },
      },
    },
    {
      description: 'Posts ordered by the last name of the author',
      query: {
        collectionName: 'posts',
        order: [
          [
            'author.name',
            'ASC',
            {
              subquery: {
                collectionName: 'users',
                where: [['id', '=', '$1.authorId']],
              },
              cardinality: 'one',
            },
          ],
        ],
      },
    },
  ];

  test.each(queries)('query: $description', async ({ query }) => {
    const plan = compileQuery(query, undefined);
    expect(plan).toMatchSnapshot();
  });
});

function printInAsciiBox(text: string) {
  const lines = text.split('\n');
  const maxLength = Math.max(...lines.map((line) => line.length));
  const horizontalBorder = '─'.repeat(maxLength + 2);
  console.log(`┌${horizontalBorder}┐`);
  for (const line of lines) {
    console.log(`│ ${line.padEnd(maxLength)} │`);
  }
  console.log(`└${horizontalBorder}┘`);
}
