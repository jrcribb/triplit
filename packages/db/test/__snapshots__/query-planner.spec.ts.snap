// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`query planning > query: '20 most recent messages in a conversa…' 1`] = `
{
  "steps": [
    {
      "collectionName": "messages",
      "type": "SCAN",
    },
    {
      "filter": [
        [
          "conversationId",
          "=",
          "conv123",
        ],
      ],
      "type": "ITERATOR_FILTER",
    },
    {
      "type": "COLLECT",
    },
    {
      "fields": [
        [
          "timestamp",
          "DESC",
        ],
      ],
      "type": "SORT",
    },
    {
      "count": 20,
      "type": "LIMIT",
    },
  ],
  "views": {},
}
`;

exports[`query planning > query: 'Benchmarks query' 1`] = `
{
  "steps": [
    {
      "collectionName": "branches",
      "type": "SCAN",
    },
    {
      "type": "COLLECT",
    },
    {
      "alias": "latest_run",
      "subPlan": [
        {
          "type": "PREPARE_VIEW",
          "viewId": "0",
        },
        {
          "filter": [
            [
              "branch_name",
              "=",
              "$1.id",
            ],
          ],
          "type": "RESOLVE_FROM_VIEW",
          "viewId": "0",
        },
        {
          "type": "PICK",
        },
      ],
      "type": "SUBQUERY",
    },
    {
      "alias": "benchmarks",
      "subPlan": [
        {
          "collectionName": "benchmarks",
          "type": "SCAN",
        },
        {
          "type": "COLLECT",
        },
        {
          "alias": "latest_branch_run",
          "subPlan": [
            {
              "type": "PREPARE_VIEW",
              "viewId": "1",
            },
            {
              "filter": [
                [
                  "benchmark",
                  "=",
                  "$1.id",
                ],
                [
                  "branch_name",
                  "=",
                  "$2.id",
                ],
              ],
              "type": "RESOLVE_FROM_VIEW",
              "viewId": "1",
            },
            {
              "type": "PICK",
            },
          ],
          "type": "SUBQUERY",
        },
      ],
      "type": "SUBQUERY",
    },
  ],
  "views": {
    "0": {
      "steps": [
        {
          "collectionName": "runs",
          "type": "SCAN",
        },
        {
          "type": "COLLECT",
        },
        {
          "fields": [
            [
              "branch_name",
              "ASC",
            ],
            [
              "created_at",
              "DESC",
            ],
          ],
          "type": "SORT",
        },
      ],
      "views": {},
    },
    "1": {
      "steps": [
        {
          "collectionName": "runs",
          "type": "SCAN",
        },
        {
          "type": "COLLECT",
        },
        {
          "fields": [
            [
              "benchmark",
              "ASC",
            ],
            [
              "branch_name",
              "ASC",
            ],
            [
              "created_at",
              "DESC",
            ],
          ],
          "type": "SORT",
        },
      ],
      "views": {},
    },
  },
}
`;

exports[`query planning > query: 'Conversations including the most rece…' 1`] = `
{
  "steps": [
    {
      "collectionName": "conversations",
      "type": "SCAN",
    },
    {
      "type": "COLLECT",
    },
    {
      "alias": "latestMessage",
      "subPlan": [
        {
          "collectionName": "messages",
          "type": "SCAN",
        },
        {
          "filter": [
            [
              "conversationId",
              "=",
              "$1.id",
            ],
          ],
          "type": "ITERATOR_FILTER",
        },
        {
          "type": "COLLECT",
        },
        {
          "fields": [
            [
              "timestamp",
              "DESC",
            ],
          ],
          "type": "SORT",
        },
        {
          "alias": "sender",
          "subPlan": [
            {
              "type": "PREPARE_VIEW",
              "viewId": "0",
            },
            {
              "filter": [
                [
                  "id",
                  "=",
                  "$1.senderId",
                ],
              ],
              "type": "RESOLVE_FROM_VIEW",
              "viewId": "0",
            },
            {
              "type": "PICK",
            },
          ],
          "type": "SUBQUERY",
        },
        {
          "type": "PICK",
        },
      ],
      "type": "SUBQUERY",
    },
  ],
  "views": {
    "0": {
      "steps": [
        {
          "collectionName": "users",
          "type": "SCAN",
        },
        {
          "type": "COLLECT",
        },
        {
          "fields": [
            [
              "id",
              "ASC",
            ],
          ],
          "type": "SORT",
        },
      ],
      "views": {},
    },
  },
}
`;

exports[`query planning > query: 'Flights to california including the d…' 1`] = `
{
  "steps": [
    {
      "collectionName": "flights",
      "type": "SCAN",
    },
    {
      "type": "PREPARE_VIEW",
      "viewId": "0",
    },
    {
      "filter": [
        [
          "destination",
          "in",
          "$view_0.id",
        ],
      ],
      "type": "ITERATOR_FILTER",
    },
    {
      "type": "COLLECT",
    },
    {
      "alias": "destinationAirport",
      "subPlan": [
        {
          "type": "PREPARE_VIEW",
          "viewId": "1",
        },
        {
          "filter": [
            [
              "id",
              "=",
              "$1.destination",
            ],
          ],
          "type": "RESOLVE_FROM_VIEW",
          "viewId": "1",
        },
        {
          "type": "PICK",
        },
      ],
      "type": "SUBQUERY",
    },
  ],
  "views": {
    "0": {
      "steps": [
        {
          "collectionName": "airports",
          "type": "SCAN",
        },
        {
          "filter": [
            [
              "state",
              "=",
              "CA",
            ],
          ],
          "type": "ITERATOR_FILTER",
        },
        {
          "type": "COLLECT",
        },
      ],
      "views": {},
    },
    "1": {
      "steps": [
        {
          "collectionName": "airports",
          "type": "SCAN",
        },
        {
          "filter": [
            [
              "state",
              "=",
              "CA",
            ],
          ],
          "type": "ITERATOR_FILTER",
        },
        {
          "type": "COLLECT",
        },
        {
          "fields": [
            [
              "id",
              "ASC",
            ],
          ],
          "type": "SORT",
        },
      ],
      "views": {},
    },
  },
}
`;

exports[`query planning > query: 'Messages and the reactions' 1`] = `
{
  "steps": [
    {
      "collectionName": "messages",
      "type": "SCAN",
    },
    {
      "type": "COLLECT",
    },
    {
      "alias": "reactions",
      "subPlan": [
        {
          "type": "PREPARE_VIEW",
          "viewId": "0",
        },
        {
          "filter": [
            [
              "messageId",
              "=",
              "$1.id",
            ],
          ],
          "type": "RESOLVE_FROM_VIEW",
          "viewId": "0",
        },
      ],
      "type": "SUBQUERY",
    },
  ],
  "views": {
    "0": {
      "steps": [
        {
          "collectionName": "reactions",
          "type": "SCAN",
        },
        {
          "type": "COLLECT",
        },
        {
          "fields": [
            [
              "messageId",
              "ASC",
            ],
          ],
          "type": "SORT",
        },
      ],
      "views": {},
    },
  },
}
`;

exports[`query planning > query: 'Posts authored by user with username …' 1`] = `
{
  "steps": [
    {
      "collectionName": "posts",
      "type": "SCAN",
    },
    {
      "type": "PREPARE_VIEW",
      "viewId": "0",
    },
    {
      "filter": [
        [
          "authorId",
          "in",
          "$view_0.id",
        ],
      ],
      "type": "ITERATOR_FILTER",
    },
    {
      "type": "COLLECT",
    },
  ],
  "views": {
    "0": {
      "steps": [
        {
          "collectionName": "users",
          "type": "SCAN",
        },
        {
          "filter": [
            [
              "username",
              "=",
              "Bob99",
            ],
          ],
          "type": "ITERATOR_FILTER",
        },
        {
          "type": "COLLECT",
        },
      ],
      "views": {},
    },
  },
}
`;

exports[`query planning > query: 'Posts ordered by the last name of the…' 1`] = `
{
  "steps": [
    {
      "collectionName": "posts",
      "type": "SCAN",
    },
    {
      "type": "COLLECT",
    },
    {
      "alias": "_order_0",
      "subPlan": [
        {
          "type": "PREPARE_VIEW",
          "viewId": "0",
        },
        {
          "filter": [
            [
              "id",
              "=",
              "$1.authorId",
            ],
          ],
          "type": "RESOLVE_FROM_VIEW",
          "viewId": "0",
        },
      ],
      "type": "SUBQUERY",
    },
    {
      "fields": [
        [
          "_order_0.name",
          "ASC",
        ],
      ],
      "type": "SORT",
    },
  ],
  "views": {
    "0": {
      "steps": [
        {
          "collectionName": "users",
          "type": "SCAN",
        },
        {
          "type": "COLLECT",
        },
        {
          "fields": [
            [
              "id",
              "ASC",
            ],
          ],
          "type": "SORT",
        },
      ],
      "views": {},
    },
  },
}
`;

exports[`query planning > query: 'messages from conversations that the …' 1`] = `
{
  "steps": [
    {
      "collectionName": "messages",
      "type": "SCAN",
    },
    {
      "type": "PREPARE_VIEW",
      "viewId": "0",
    },
    {
      "filter": [
        [
          "conversationId",
          "in",
          "$view_0.id",
        ],
      ],
      "type": "ITERATOR_FILTER",
    },
    {
      "type": "COLLECT",
    },
  ],
  "views": {
    "0": {
      "steps": [
        {
          "type": "PREPARE_VIEW",
          "viewId": "1",
        },
        {
          "collectionName": "conversations",
          "ids": "$view_1.conversationId",
          "type": "ID_LOOK_UP",
        },
        {
          "type": "COLLECT",
        },
      ],
      "views": {},
    },
    "1": {
      "steps": [
        {
          "collectionName": "conversation_members",
          "type": "SCAN",
        },
        {
          "filter": [
            [
              "userId",
              "=",
              "user123",
            ],
          ],
          "type": "ITERATOR_FILTER",
        },
        {
          "type": "COLLECT",
        },
      ],
      "views": {},
    },
  },
}
`;
